velox/vector/BaseVector.cpp

```diff
@@ -785,5 +819,69 @@ void BaseVector::prepareForReuse() {
   }
 }
 
+namespace {
+
+size_t typeSize(const Type& type) {
+  switch (type.kind()) {
+    case TypeKind::VARCHAR:
+    case TypeKind::VARBINARY:
+      return sizeof(StringView);
+    case TypeKind::OPAQUE:
+      return sizeof(std::shared_ptr<void>);
+    default:
+      VELOX_DCHECK(type.isPrimitiveType(), type.toString());
+      return type.cppSizeInBytes();
+  }
+}
+
+struct BufferReleaser {
+  explicit BufferReleaser(const BufferPtr& parent) : parent_(parent) {}
+  void addRef() const {}
+  void release() const {}
+
+ private:
+  BufferPtr parent_;
+};
+
+BufferPtr sliceBufferZeroCopy(
+    size_t typeSize,
+    bool podType,
+    const BufferPtr& buf,
+    vector_size_t offset,
+    vector_size_t length) {
+  // Cannot use `Buffer::as<uint8_t>()` here because Buffer::podType_ is false
+  // when type is OPAQUE.
+  auto data =
+      reinterpret_cast<const uint8_t*>(buf->as<void>()) + offset * typeSize;
+  return BufferView<BufferReleaser>::create(
+      data, length * typeSize, BufferReleaser(buf), podType);
+}
+
+} // namespace
+
+// static
+BufferPtr BaseVector::sliceBuffer(
+    const Type& type,
+    const BufferPtr& buf,
+    vector_size_t offset,
+    vector_size_t length,
+    memory::MemoryPool* pool) {
+  if (!buf) {
+    return nullptr;
+  }
+  if (type.kind() != TypeKind::BOOLEAN) {
+    return sliceBufferZeroCopy(
+        typeSize(type), type.isPrimitiveType(), buf, offset, length);
+  }
+  if (offset % 8 == 0) {
+    return sliceBufferZeroCopy(1, true, buf, offset / 8, (length + 7) / 8);
```

Comment:
nit: why do we add length + 7 here ? Is the idea to do ceil(length / 8) ?

Reply:
Yes it's a more precise and efficient way to do `ceil(length / 8)` on integers